text/0000-build-std.md

```diff
@@ -0,0 +1,4582 @@
+[related-work]: #related-work
+
+There are a variety of ongoing efforts, ideas, RFCs or draft notes describing
+features that are related or would be beneficial for build-std:
+
+- **[Opaque dependencies]**, [epage], May 2025
+  - Introduces the concept of an opaque dependency that has its own
+    `Cargo.lock`, `RUSTFLAGS` and `profile`
+  - Opaque dependencies could enable a variety of build-time performance
+      improvements:
+    - Caching - differences in dependency versions can cause unique instances of
+      every dependent crate
+    - Pre-built binaries - can leverage a pre-built artifact for a given opaque
+      dependency
+      - e.g. the standard library's distributed `rlib`s
+    - MIR-only/cross-crate lazy compilation - Small dependencies could be built
+      lazily and larger dependencies built once
+    - Optimising dependencies - dependencies could always be optimised when they
+      are unlikely to be needed during debugging
+
+# Motivation
+[motivation]: #motivation
+
+While the pre-built standard library has been sufficient for the majority of
+Rust users, there are a variety of use-cases which require the ability to
+re-build the standard library.
+
+This RFC aims to support the following use cases:
+
+1. **Re-building the standard library with different codegen flags or profile**
+   ([wg-cargo-std-aware#2])
+
+ - Embedded users need to optimise aggressively for size, due to the limited
+   space available on their target platforms, which can be achieved in Cargo by
+   setting `opt-level = s/z` and `panic = "abort"` in their profile. However,
+   these settings will not apply to the pre-built standard library
+ - Similarly, when deploying to known environments, use of `target-cpu` or
+   `target-feature` can improve the performance of code generation or allow the
+   use of newer hardware features than the target's baseline provides. As above,
+   these configuration will not apply to the pre-built standard library
+ - While the pre-built standard library is built to support debugging without
+   compromising size and performance by setting `debuginfo=1`, this isn't
+   ideal, and building the standard library with the dev profile would provide
+   a better experience
+
+2. **Unblock stabilisation of ABI-modifying compiler flags**
+
+  - Any compiler flags which change the ABI cannot currently be stabilised as they
+    would immediately mismatch with the pre-built standard library
+    - Without an ability to rebuild the standard library using these flags, it is
+      impossible to use them effectively and safely if stabilised
+  - ABI-modifying flags are designated as target modifiers ([rfcs#3716]/[rust#136966])
+    and require that the same value for the flag is passed to all compilation units
+    - Flags which need to be set across the entire crate graph to uphold some
+      property (i.e. enhanced security) are also target modifiers
+    - For example: sanitizers, control flow integriy, `-Zfixed-x18`, etc
```

Comment:
Question: How important/negotiable are these goals if they will significantly increase the contentiousness of this proposal?

Stabilizing ABI-modifying flags or otherwise custom build options, as described here, might lead to problems regarding what guarantees, especially the libs team can provide. It's hard to know nor test that all possible combinations of build options/flags/targets work, and it seems like a particularly hard goal.

Making this available on nightly is at least more realistic. But I feel like if we wanted to ship anything to stable, it might be better to only focus on building tier-3 targets for a stable MVP.